File: csharp/SetsGameSolver/SetsSolutionEngine/Solver.cs
using System.Collections.Generic;
using System.Linq;

namespace pauloq.SetsSolutionEngine
{
    public class Solver
    {
        public Board MyBoard { get; }

        public Solver() : this(new Board())
        {
        }

        public Solver(Board board)
        {
            MyBoard = board;
        }

        public static bool IsSet(IEnumerable<Card> cards)
        {
            var cardCount = cards.Count();
            var colorCount = cards.Select(x => x.Color).Distinct().Count();
            var shapeCount = cards.Select(x => x.Shape).Distinct().Count();
            var fillCount = cards.Select(x => x.Fill).Distinct().Count();
            var numberCount = cards.Select(x => x.Number).Distinct().Count();

            return ((colorCount == cardCount || colorCount == 1)
                && (shapeCount == cardCount || shapeCount == 1)
                && (fillCount == cardCount || fillCount == 1)
                && (numberCount == cardCount || numberCount == 1));
        }

        public IEnumerable<IEnumerable<Card>> BruteForceSolve()
        {
            var cards = MyBoard.Cards;
            int i = 0;

            foreach (var card1 in cards)
            {
                i++;
                var j = 0;

                foreach (var card2 in cards.Skip(i))
                {
                    j++;

                    if (card1.Equals(card2)) continue;

                    foreach (var card3 in cards.Skip(j))
                    {
                        if (card1.Equals(card3) || card2.Equals(card3)) continue;
                        var set = new[] { card1, card2, card3 };

                        if (IsSet(set))
                        {
                            yield return set;
                        }
                    }

                }
    
            }
        }
    }
}

File: golang/setsgame/solver.go
package setsgame

func IsSet(cards []Card) bool {
	distinctColors := make(map[Color]bool)
	distinctShapes := make(map[Shape]bool)
	distinctFills := make(map[Fill]bool)
	distinctSizes := make(map[Size]bool)

	for _, card := range cards {
		distinctColors[card.Color] = true
		distinctShapes[card.Shape] = true
		distinctFills[card.Fill] = true
		distinctSizes[card.Size] = true
	}

	colorCount := len(distinctColors)
	shapeCount := len(distinctShapes)
	fillCount := len(distinctFills)
	sizeCount := len(distinctSizes)

	return ((colorCount == 3 || colorCount == 1) &&
		(shapeCount == 3 || shapeCount == 1) &&
		(fillCount == 3 || fillCount == 1) &&
		(sizeCount == 3 || sizeCount == 1))
}

func BruteForceSolve(board *Board) [][]Card {
	result := make([][]Card, 0, 6)
	i := 0

	for _, card1 := range board.Cards {
		i++
		j := i

		for _, card2 := range board.Cards[i:] {
			j++

			for _, card3 := range board.Cards[j:] {
				set := []Card{card1, card2, card3}

				if IsSet(set) {
					result = append(result, set)
				}
			}
		}
	}

	return result
}
